Service pipelines using variable groups from bedrock.yaml

sarath-p opened the pull request:
- Closes https://github.com/microsoft/bedrock/issues/887
- Checks if user ran `spk project init` and `spk project cvg` commands before this `spk service create`

dennisseah commented:
We shall have conflict with other PR that I have https://github.com/CatalystCode/spk/pull/215
I have moved isBedrockFileExist to a new helper file where it contains all the bedrock related functions.

sarath-p commented:
> We shall have conflict with other PR that I have #215
> I have moved isBedrockFileExist to a new helper file where it contains all the bedrock related functions.

The function code also changed. We can merge through master branch.

dennisseah commented on src/config.ts:
```diff
@@ -294,3 +294,31 @@ export const saveConfiguration = async (
+
+  logger.debug(`variableGroups length: ${bedrockFile?.variableGroups?.length}`);
+  logger.verbose(`bedrockFile: \n ${bedrockFile}`);
+  return [true, bedrockFile?.variableGroups?.length];
```
you are overloading this function. I hope that you can return `true` or `false` here. and then do variableGroups length check in the caller function. When we overload function, you can see that you needed to return things like undefined. thanks.

sarath-p replied:
It is a deep check i.e. not only a file exist, but need to make sure variable groups exist in the file. In this scenario, the caller does't need the file contents and true/false is sufficient. Perhaps we can change the second one to `boolean` and remove the `undefined`.  Let me know your thoughts on this.

dennisseah replied:
I am misled by the function name, isBedrockFileValid. As the function name suggest, I thought that we are checking if bedrock file exist or not :-) I think we can do this.
```
interface IBedrockFileStat {
  exist: boolean;
  hasVariableGroups: boolean;
}

export const bedrockFileStat = async (
  rootProjectPath: string
): Promise<IBedrockFileStat> => {
  if (!rootProjectPath) {
    throw new Error("Project root path is not valid");
  }

  const absProjectPath = path.resolve(rootProjectPath);

  let bedrockFile: IBedrockFile | undefined;

  try {
    const bedrockFile = await BedrockAsync(absProjectPath);
    return {
      exist: true,
      hasVariableGroups: bedrockFile?.variableGroups?.length
    }
  } catch (error) {
    logger.error(error);
    return {
      exist: false,
    }
  }
}
```

this enables other team members to reuse and enhance this function get more stats from bedrock file.
and we do not return array, we return a well defined interface.

what do you think?

sarath-p replied:
Yes, the name was misleading. I thought about it and am going to refactor it to check whether dependency commands were executed.  One of the goal it to centralize the error messages too so they won't be repeated.

sarath-p replied:
Discussed with @mtarng and the decided to keep it as `bedrock file info` and use it to validate the project commands order.

dennisseah replied:
sorry, I do not understand `keep it as bedrock file info`. Do you mind elaborate?

sarath-p replied:
I was thinking to implement command sequence validator using this logic, but decided to keep the method name as bedrock file that may be used for other checks in the future.

dennisseah replied:
ok

mtarng commented on src/commands/project/create-variable-group.ts:
```diff
@@ -76,72 +82,80 @@ export const execute = async (
-      logger.error(`Error occurred while creating variable group`);
-      logger.error(err);
-      await exitFn(1);
+    const bedrockFileIno = await bedrockFileInfo(projectPath);
```
Typo?
>bedrockFileIno

or bedrockFileInfo

sarath-p replied:
yes, let me fix it
